### PR TITLE
Properly throw errors from SubmitResult

### PR DESCRIPTION
**Pull request by birchmd**
Should make it so that revert messages are reported by eth tooling (eg remix)

### CODE REVIEW

File: lib/errors.d.ts [hunk 1]
```diff
@@ -34,5 +34,5 @@ export declare class TransactionError extends ExpectedError {
     constructor(message: string);
 }
 export declare class RevertError extends ExpectedError {
-    constructor(reason: Uint8Array);
```
[the remark this answers is not in the document]
Comment (0x3bfc) [reply lib/errors.d.ts:37]: @birchmd  noted and working on it.

**Comment by Sanmyohtet9** [comment]
<∆>*,</∆>

File: lib/errors.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (artob) [reply lib/errors.d.ts:37]: @0x3bfc What's the status of this?

[the remark this answers is not in the document]
Comment (artob) [reply lib/errors.d.ts]: @0x3bfc Is that from `revert("Error from My reverter")`?

[the remark this answers is not in the document]
Comment (0x3bfc) [reply lib/errors.d.ts]: Yes, I am expected only printable error message. Example below:

```javascript
contract Reverter {
    address data = address(0);

    function myRevertFunction() external returns (bool) {
        require(
            msg.sender == data,
            'Error from My reverter'
        );
        data = msg.sender;
        return true;
    }
}
```

[the remark this answers is not in the document]
Comment (artob) [reply lib/errors.d.ts]: Were you testing with this branch, or the current `master` branch? I assume the binary content here is a Borsh-encoded envelope.

[the remark this answers is not in the document]
Comment (0x3bfc) [reply lib/errors.d.ts]: No this branch, but I already merged the master to this one. 
> I assume the binary content here is a Borsh-encoded envelope.
will check it out.

[the remark this answers is not in the document]
Comment (birchmd) [reply lib/errors.d.ts]: The data is ABI encoded with a constant byte sequence out front. See https://github.com/ethereum/solidity/pull/3364 for more details.

In rust the code to parse out the bytes looks like
```rust
let revert_bytes = crate::test_utils::unwrap_revert(submit_result);
// First 4 bytes is a function selector with signature `Error(string)`
assert_eq!(&revert_bytes[0..4], &[8, 195, 121, 160]);
// Remaining data is an ABI-encoded string
let revert_message = ethabi::decode(&[ethabi::ParamType::String], &revert_bytes[4..]).unwrap();
println!("{:?}", revert_message);
```

[the remark this answers is not in the document]
Comment (artob) [reply lib/errors.d.ts]: @0x3bfc This seems a good explanation: https://ethereum.stackexchange.com/a/66173/

And for implementations there's [a simple one](https://gist.github.com/gluk64/fdea559472d957f1138ed93bcbc6f78a) and then [an NPM package](https://www.npmjs.com/package/eth-revert-reason) which may or may not be usable with current Ethers.js releases.

[the remark this answers is not in the document]
Comment (0x3bfc) [reply lib/errors.d.ts]: I tested it with and without a revert message and both are working.

**Review by artob — APPROVED** [review]

**Pull request merged**